### ISSUE TITLE
Unable to perform segmentation

### ISSUE THREAD

**Issue by soumickmj** [mazurowski-lab/mri_foundation#3]
Hello,
Thanks for sharing this amazing work!

I'm trying to segment MRI abdominal slices using bounding boxes and point prompts, as I did with MedSAM earlier.
Following the readme file of this repo, after downloading the weights and loading the model, I did the following:

`
   with torch.no_grad():

       image_embeddings = model.image_encoder(preprocessed_img)

       sparse_embeddings, dense_embeddings = model.prompt_encoder(points=None, boxes=bbox_torch, masks=None)

       pred, _ = model.mask_decoder(
            image_embeddings=image_embeddings,
            image_pe=model.prompt_encoder.get_dense_pe(),
            sparse_prompt_embeddings=sparse_embeddings,
            dense_prompt_embeddings=dense_embeddings,
            multimask_output=False,
        )
`

I also tried the prompt-free one (auto mode):
`
  with torch.no_grad():

       image_embeddings = model.image_encoder(preprocessed_img)

       sparse_embeddings, dense_embeddings = model.prompt_encoder(points=None, boxes=None, masks=None)

       pred, _ = model.mask_decoder(
            image_embeddings=image_embeddings,
            image_pe=model.prompt_encoder.get_dense_pe(),
            sparse_prompt_embeddings=sparse_embeddings,
            dense_prompt_embeddings=dense_embeddings,
            multimask_output=False,
        )
`
and got similar issues!

I was expecting this pred to be my segmentation, but it isn't; rather, it looks like some kind of structured/patterned noise. 

Am I missing something here? Maybe I need to do some kind of fine-tuning for my task?
It would be really helpful if you could create a small demo to show how the different downstream task workflows are (including segmentation).

Thank you so much!

**Comment by haoyudong-97** [comment]
Thank you for the interest in our work! We only pre-train the image encoder with self-supervised learning, which means it does not support prompt-based segmentation. To support it you need to further fine-tune it with some datasets.

**Comment by soumickmj** [comment]
@haoyudong-97 thanks for your response!
I understand the prompt-based isn't working, but I'm confused about the auto-mode.

I did the following:
`with torch.no_grad():
image_embeddings = model.image_encoder(preprocessed_img)

   sparse_embeddings, dense_embeddings = model.prompt_encoder(points=None, boxes=None, masks=None)

   pred, _ = model.mask_decoder(
        image_embeddings=image_embeddings,
        image_pe=model.prompt_encoder.get_dense_pe(),
        sparse_prompt_embeddings=sparse_embeddings,
        dense_prompt_embeddings=dense_embeddings,
        multimask_output=False,
    )`

In this case, pred should be the segmentation generated in auto-mode, right?
But, the output I'm getting doen't make sense, it's like small tiny objects or artefacts all over the output.
Could you please let me know with some code snippets how can I use it for segmentation? (if I'm doing something wrong).

Thanks!

**Comment by haoyudong-97** [comment]
Ah I see your question. No, our model does not support auto-mode either since we only pre-train the image encoder with extra MRI; this means the decoder cannot recognize image encoder's output without extra fine-tuning. 

Though one of the core contributions of our paper is that you should get good segmentation results by fine-tuning it only on a few MRI slices.

**Comment by soumickmj** [comment]
Ok, thank you so much for your clarification.

One follow-up question!
Is there any plans to publish the downstream codes / demos (or they are already published)? If yes, could you kindly let me know what is the expected timeline?

Thank you so much for your help!

**Comment by haoyudong-97** [comment]
A lot of the downstream codes for segmentation are from another repo of our lab: https://github.com/mazurowski-lab/finetune-SAM. 

We may release the downstream codes for other tasks when I become more available :)
